Add explanation/concept for extension maturity model (#42446)
quarkusio/quarkus: docs/src/main/asciidoc/extension-maturity-model.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="extension-maturity-model"]
= A maturity model for Quarkus extensions
include::_attributes.adoc[]
:diataxis-type: concept
:categories: writing-extensions
:topics: extensions
:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors might want to include.
////
The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
////

What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?

> image::extension-maturity-model.svg[A maturity model: Runtime actually works,Works in JVM mode,Works in dev mode, Works in native,Configuration support,Dev service,Dev UI, Use build-time application knowledge to eliminate wasteful runtime code paths (supersonic subatomic performance),Use build-time application knowledge to remove boilerplate,Codestart application template]

Reviewer: what does the bracketed smiley face indicate?

Author: Developer joy. So the idea is that there are several different things extension owners should do in the theme of "developer joy". Now that we have one, we could also take the "performance" icon from https://github.com/quarkusio/quarkusio.github.io/pull/2122 and put that on the performance-related items.

Reviewer: It took me way more time than I care to admit that that icon is the same as the one shown on https://quarkus.io/developer-joy/ :)

Could we somehow link/highlight that - to somehow guide daft readers (like me :) to make the connection.

Author: I'm hoping it will be a lot more clear once we have a performance icon we can annotate the performance items with, so then people will recognise two items from the front page - but that needs https://github.com/quarkusio/quarkusio.github.io/pull/2122 to define a performance icon, and then it needs the front page updates, and all that work is still in progress. So I don't think it makes sense to block this content on that work.

Author: I'd suggested focussing on the 'what', not the 'how', to tie it back to the front page. (Which also should focus on the benefits to users, not the 'how'.) So categories should be something like

 - developer joy - dev UI, hot reload, dev services, unified config, boilerplate reduction, programming model simplification
-  performance - goes fast, uses very few resources (and the ingredients of this is the reactive core and the build first principle, but those are how, not what)
 -  prod/cloud readiness - observability, manageability, operations concerns

image::extension-maturity-model.svg[A maturity model: Runtime actually works,Works in JVM mode,Works in dev mode, Works in native,Configuration support,Dev service,Dev UI, Use build-time application knowledge to eliminate wasteful runtime code paths (supersonic subatomic performance),Use build-time application knowledge to remove boilerplate,Codestart application template]

The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!
Also, not every step will apply to every extension. For example, you don't need a Dev Service if your extension doesn't depend on external services.
It's completely OK to publish a first version of an extension that doesn't handle everything. In fact, it's OK if your extension _never_ gets to the more advanced features. This is a suggested pathway, not a minimum feature set.

Also note that this list only includes the technical features of your extension.
You might also want to think about how you share your extension, and how it presents itself to the world.

> The link:https://hub.quarkiverse.io/checklistfornewprojects/[new extension checklist] on the Quarkiverse Hub has a useful list of ways extensions can participate in the ecosystem.

Reviewer: not blocking this pr for it but that link renders really badly for me. 

And the links on left on that page seem to be missing a lot compared to https://github.com/quarkiverse/quarkiverse/wiki/ChecklistForNewProjects - are these two using the same sources or split somehow?

Author: They're using the same source. The rendered version has the significant advantage that it's indexed by google, and thus exists to people trying to find the content. It's also got the advantages that the hosting isn't tied to github wikis, and we have a lot of control over the appearance (we just haven't invested the time to take advantage of that control).  

The navigation depth is different in the two renders, and I see there's some page width issues on a few rows. Is that the rendering issues you mean? Want to raise a defect against https://github.com/quarkiverse/quarkiverse for what you're seeing?

Reviewer: sure - done, https://github.com/quarkiverse/quarkiverse/issues/288

=== Works in dev mode

In some cases, extra work may be needed to ensure any wrapped libraries can tolerate
dev mode, since the classloading is different and hot reloading can break some assumptions. Extensions may also wish to add some
xref:writing-extensions.adoc#integrating-with-development-mode[special handling for dev mode].
To add automated tests which validate dev mode, you can xref:writing-extensions.adoc#testing-hot-reload[add tests which extend the `QuarkusDevModeTest`].

=== Works as a native application

For many libraries, native mode support is the primary motivation for creating an extension. See xref:writing-extensions.adoc#native-executable-support[the guide on native executable support] for more discussion about some of the adaptations that might be needed.

== Configuration support

Extensions should support Quarkus's unified configuration, by xref:writing-extensions.adoc#configuration[integrating with the Quarkus configuration model].
The Writing Extensions guide has more guidance on xref:writing-extensions.adoc#how-to-expose-configuration[the Quarkus configuration philosophy].

> == Dev service (if there is an external service)
> 
> To provide a Dev Service, use the `DevServicesResultBuildItem` build item. See the xref:extension-writing-dev-service.adoc[Dev Services how-to] for more information.

Reviewer: ```suggestion
== Dev service

Dev services are generally relevant for extensions that "connects" to something. i.e. databases for datasources, keycloak instance for security, Apache Kafka instance for kafka, etc.

If relevant for your extension, it can provide a Dev Service, use the `DevServicesResultBuildItem` build item. See the xref:extension-writing-dev-service.adoc[Dev Services how-to] for more information.
```

Just a suggestion as I think many will miss the "(if there is an external service)" in the header.

Author: Let's do both, and keep the header plus add the text.

== Dev UI

Extensions can take advantage of the Dev UI to present extra information to users. See the xref:dev-ui.adoc[Dev UI for extension developers] guide.

> == Use build-time application knowledge to remove boilerplate

Reviewer: should this have the title "Joyful programming model" to link back to the staircase?

Reviewer: also this and supersonic performance shuold move up when/if doing the suggested reordering as mentioned above.

Author: I've updated the heading. I originally had it longer because there was the space and "Use build-time application knowledge to remove boilerplate" communicates more than "Joyful programming model," but I'll add that to the text so the link to the staircase is clear.

> 

Reviewer 2: No empty line for the header and please add the top comment:

```
////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
```

This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="extension-maturity-model"]
= A maturity model for Quarkus extensions
include::_attributes.adoc[]
:diataxis-type: concept
:categories: writing-extensions
:topics: extensions
:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors might want to include.
////
The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
////

> What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?

Author: Good point, added.

** Works in native
* Configuration support
* Dev service (if there is an external service dependency)
* Dev UI
* Use build-time application knowledge to eliminate wasteful runtime code paths (supersonic subatomic performance)
* Use build-time application knowledge to remove boilerplate
* Codestart application template

The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!

> Also, not every step will apply to every extension. For example, you don't need a dev service if your extension doesn't depend on external services.

Reviewer 2: I think it needs to be clear that you don't need to implement all of this before releasing your extension. It's more or less mentioned but people have a tendency to go over the top sometimes when developing in the open and we need to clarify it's OK to publish a first version that doesn't handle everything.

Reviewer 2: Perfect! (sorry forgot about it!)

It's also a good idea to spend some time on the metadata in the link:https://quarkus.io/version/main/guides/extension-metadata#quarkus-extension-yaml[`quarkus-extension.yaml` file], which is used by Quarkus tooling.

Here are some pointers on how to achieve those capabilities.

== Runtime actually works

=== Works in JVM mode

For most extensions, this is the minimum expectation.
When wrapping an existing library, this is usually trivial to achieve; if an extension is providing net-new capability, it might be a bit more work. Quarkus provides tools for xref:writing-extensions.adoc#testing-extensions[unit testing and integration testing] extensions.

> 

Reviewer 3: Here there is (at least) two categories:

- just wrapping the library
- exposing _managed_ clients

Author: What do we mean by a managed client in this context? Just injectable by CDI, or something more? Is that starting to get into the area of what I'd called programming model improvements elsewhere in the doc?

For most extensions, this is the minimum expectation.
When wrapping an existing library, this is usually trivial to achieve; if an extension is providing net-new capability, it might be a bit more work. Quarkus provides tools for xref:writing-extensions.adoc#testing-extensions[unit testing and integration testing] extensions.

=== Works in dev mode

In some cases, extra work may be needed to ensure any wrapped libraries can tolerate
dev mode, since the classloading is different and hot reloading can break some assumptions. Extensions may also wish to add some
xref:writing-extensions.adoc#integrating-with-development-mode[special handling for dev mode].
To add automated tests which validate dev mode, you can xref:writing-extensions.adoc#testing-hot-reload[add tests which extend the `QuarkusDevModeTest`].

=== Works as a native application

> For many libraries, native mode support is the motivation for creating an extension. See xref:writing-extensions.adoc#native-executable-support[the guide on native executable support] for more discussion about some of the adaptations that might be needed.

Reviewer 2: ```suggestion
For many libraries, native mode support is the primary motivation for creating an extension. See xref:writing-extensions.adoc#native-executable-support[the guide on native executable support] for more discussion about some of the adaptations that might be needed.
```

My opinion about this is that it's a bit sad given you can do crazy things with a nice Quarkus integration but I suppose that's true anyway :).

== Configuration support

Extensions should support Quarkus's unified configuration, by xref:writing-extensions.adoc#configuration[integrating with the Quarkus configuration model].
The Writing Extensions guide has more guidance on xref:writing-extensions.adoc#how-to-expose-configuration[the Quarkus configuration philosophy].

== Dev service

> To provide a dev service, use the `DevServicesResultBuildItem` build item.

Reviewer 2: I think we should try to add pointers to extensions that are good inspiration. Does it make sense?

If so, I suppose your next question will be for me to point them to you? :)

Author: es, and no. And yes, but no. But yes. :) For dev services in particular, I’m planning to get out a basic dev service documentation this week, and then cross-link to it from here. So ‘do it this this week’ can sometimes be a fiction, so there’s no harm having links to examples in the interim, but we also don’t need to spend too much time on stuff that I’m hoping to remove in a day or two. Those links to good examples would be helpful to include in the dev services doc, though.  More generally, I really like the idea of providing pointers to example extensions. I’ve got another doc in draft that’s an an extension FAQ with questions like “I need to remove a method” and so on, and there I 100% want to link to examples to follow. So if you have any … :)

== Dev UI

Extensions can take advantage of the Dev UI to present extra information to users. See the xref:dev-ui.adoc[Dev UI for extension developers] guide.

== Supersonic subatomic performance

Because Quarkus moves work to the build stage, Quarkus applications should have fast startup, high throughput, and low memory requirements. Performance tuning is a large subject, but extensions should use build-time application knowledge to eliminate wasteful runtime code paths at runtime.

== Use build-time application knowledge to remove boilerplate

> Develop Joy is an important part of the Quarkus philosophy.

Reviewer 2: ```suggestion
Developer Joy is an important part of the Quarkus philosophy.
```

> At a minimum, Quarkus extensions should aim to xref:writing-extensions.adoc#expose-your-components-via-cdi[exponents components via CDI], so that they can be consumed in a frictionless way by user applications. But they can do much more.

Reviewer 2: ```suggestion
At a minimum, Quarkus extensions should aim to xref:writing-extensions.adoc#expose-your-components-via-cdi[expose components via CDI], so that they can be consumed in a frictionless way by user applications. But they can do much more.
```

[id="extension-maturity-model"]
= A maturity model for Quarkus extensions
include::_attributes.adoc[]
:diataxis-type: concept
:categories: writing-extensions
:topics: extensions
:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors may wish to include.
////
The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
////

What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?

* It actually works
** Works in JVM mode
** Works in dev mode
** Works in native

> * Dev service (if there is an external service dependency)

Reviewer: shouldn't this not be under developer joy?

Author: I wanted to pull it out to the top level, because it's so important. 

For consistency I could also pull config out to the top level, and leave developer joy to be the more amorphous 'take advantage of quarkus magic to revolutionise your programming model.'

> * It actually works
> ** Works in JVM mode
> ** Works in dev mode
> ** Works in native
> * Dev service (if there is an external service dependency)
> * Developer joy
> ** Configuration support
> ** Use build-time application knowledge to remove boilerplate
> * Dev UI
> * Supersonic subatomic performance
> ** Use build-time application knowledge to eliminate wasteful runtime code paths
> * Codestart application template

Reviewer: ```suggestion
* Runtime actually works
** Works in JVM mode
** Works in native
* Supersonic subatomic performance
** Use build-time application knowledge to eliminate wasteful runtime code paths
* Developer Joy
 ** Configuration support
 ** Dev service (if there is an external service dependency)
 ** Use build-time application knowledge to remove boilerplate
 ** Dev UI
 ** Codestart application template
```

Totally optional suggestion on grouping it a little bit different as the existing one (at least to me) felt as its categories overlapped a bit too much.

Author: I think if we group too much under developer joy it starts to break the ranking. I'd say that tolerating dev mode is table stakes for an extension, but 'redefine the programming model' is a pretty advanced feature. 

What about pulling developer joy some sort of tag or 'key'. So Config and Boilerplate reduction become top level items, dev mode goes back to being in front of native mode, and we put the joy icon at the beginning or end of those lines. Or if we do boxes, the joy icon tags the box.

Then we could do a similar tagging for the others, with some sort of 'speed' icon (I don't think we actually have one, but we should?). So then people can glance at the list and see how each thing fits in. Something like this, but not looking terrible:

<img width="680" alt="image" src="https://github.com/user-attachments/assets/1d93a205-2b88-4412-8095-716e87f6181d">

Also, in my crappy mock-up I didn't bother to adjust the order of items, and I had to steal an icon since we don't have a speed one.

Author: Or something like this, if we went for a more graphical maturity model representation:
<img width="1274" alt="image" src="https://github.com/user-attachments/assets/49fd261f-342f-4b4f-a0bd-9203bea746bb">

(Again, wrong icon for performance, and I only filled out the first bit.)

Author: I've changed 'it' to 'runtime', as per your suggestion. I've (hopefully) fixed the problem with Developer Joy you pointed out by just removing all of the 'developer joy' nesting. I've not done anything yet with icons or moving to a graphical format.

> The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!

Reviewer: the list is good - but afaics the only thing that "fits" a maturity model (as in something that is progressively more) is "works in jvm" and "works in native" (too). The rest a checkbox features.

i.e. does it help to call it a extension maturity model or does it confuse more ?

Author: I wondered the same thing, because you're right that the progression is not a strict one. But I think it does make sense to talk about the maturity of an extension. In this context it's a synonym for completeness, but completeness wouldn't be helpful term to use, because it makes it sound like people aren't 'allowed' to do extensions unless they do everything. And 'checklist' has a similar problem - it doesn't have *enough* progression. 

Another way of thinking about it is that we're trying to distinguish a really good extension from an average one, but I don't think any words like 'quality' would be helpful, because there's too much judgement. This should be an aspiration list, not a shame list.

Reviewer: yeah, I don't have too strong opinion ....just that in past we kept getting stuck trying to "rank" it .

Alternative suggestion to maturity model I can think of is Capability Model: Emphasizes the capabilities or features that are completed or available.

another is "Implementation Levels: Describes the levels at which different parts of the project have been implemented." but that also "ranks"

Author: The good thing about "maturity model" is that it will be really obvious to people what it means. "This page tells me what my extension will be like when it starts, and where it could eventually end up if I continue working on it." "Capability model" or "Implementation levels" would need some words around it to try to explain to people what the purpose of the page is, because they're not familiar concepts in the same way. 

The bad thing about maturity model is that this isn't a maturity model in the strictest sense, because the progression/dependencies on lower levels isn't strict enough. And there definitely could be some debates about the relative position of a few things. 

We could change the page title to something generic like extension-capability-model but leave the in-page title as "maturity model". That way if we decide to change the the title later on we don't have a problem with the url?

Reviewer: +1 on using capability model for the URL.

I've looked at the staircase multiple times and it keeps rubbing me the wrong way that the two cornerstone capabilities Quarkus pushes/priorities is the last two steps before last.

It would be more coherent and signal of focus if we move devservice/devui higher up maturity scale as we really want programming model and performance to be early priority.

With that its:

Works In JVM -> Works in devmode -> works in native mode -> configuration -> joyful programming model -> supersonic performance -> provide a dev service -> dev UI -> codestart

Yes, we could debate additional ordering but with that at least config/joyful/perf is better balanced with respect to what implicit message the stair case inherently brings.

Reviewer 3: Ordering will be very different from one extension to another. We should not crystalize on that. 

The only ordering we have is: works in JVM -> works in native (as the opposite is rarely true)

I would use "capabilities", or "traits" as I mentioned in another comment. 

In another comment, Max shows some radar graph. While tricky to set it right, I think it is slightly better:

- no ordering 
- extensions can decide to focus on different aspects (some dev only extension do not care or performance or cloud readiness, or even native (mailpit for example))

Also, not every step will apply to every extension. For example, you don't need a dev service if your extension doesn't depend on external services.

Also note that this list only includes the technical features of your extension.
You might also want to think about how you share your extension.
The link:https://hub.quarkiverse.io/checklistfornewprojects/[new extension checklist] on the Quarkiverse Hub has a useful list of ways extensions can participate in the ecosystem.

Here are some pointers on how to achieve those capabilities.

== It actually works

=== Works in JVM mode

For most extensions, this is the minimum expectation.
When wrapping an existing library, this is usually trivial to achieve; if an extension is providing net-new capability, it might be a bit more work. Quarkus provides tools for xref:writing-extensions.adoc#testing-extensions[unit testing and integration testing] extensions.

=== Works in dev mode

> In some cases, extra work may be needed to ensure any wrapped libraries can tolerate

Reviewer: good point but wouldn't it be better fitting on dev joy?

Author: See the comment above - I think if we're suggesting an implementation order, we can't group too much. So let's get rid of 'Developer Joy' altogether from that list, so there's no question of things being sub-points under it or not. 

Developer Joy should be a theme, not an implementation order.

Reviewer: i see the graph now have "joy full programming model" is that a missing removal or intentional?

Author: The graphic has "joyful programming model", which is intentional. "Developer Joy" includes a whole bunch of things, which would be implemented at different times - so what the graphic is talking about is specifically the boilerplate reduction, and not things like support for live reload or unified configuration.

Reviewer: ack - I like the use of the developer joy icon. its a good fit.

staircase vs graph vs something else is for sure tricky.

Its fine and at least sends a message.

Only crazy thought I had is to make it a circle as in iteratively keep improving them but it probably gets too big :)

I'm not an artist so got some help from chatgpt:

![image](https://github.com/user-attachments/assets/6728e26a-607e-475d-9d4c-cb8e528e12ef)

I kind a like this and maybe that kind of diagram could be used as  a way to express what is explained in the paragraph that "This is a suggested pathway, not a minimum feature set."

Just a thought that came to me re-reading it. Does not invalidate the simplicity of the staircase.

if it resonate we can consider it in an update.

Reviewer 3: I like the radar idea. I would have used different axis to not invent something else (and makes it less intimidating). 

Also some of them are boolean (codestart, you have it or not, how can you have half of a codestart (I have one but it does not compile?) - these should be incorporated into the "dev joy" axis.
